### ISSUE TITLE
Chain repair corrupts the freezer->leveldb transition

### ISSUE THREAD

**Issue by karalabe** [ethereum/go-ethereum#20238]
If Geth crashes, losing it's recent state, it will attempt to repair it's chain by rolling back to a block that has the state available. If however that point is in the freezer, Geth will delete the tail of the freezer **without** deleting the stuff in leveldb. This is an issue, because there will be a block gap in our dataset.

The downloader's ancestor lookup will go belly up, as it will find a recent block as available, try to import from there, which will trigger a sidechain import (parent state missing), which will hit the missing blocks at the tail of the freezer and error with unknown parent.

TL;DR: We should fix our blockchain `loadLastState` to not only set the head back, but also to include all the freezer and leveldb deletion inside. That way we know that we won't end up with junk in the database. Ideally, we might want to do this via a leveldb iterator to make sure we nuke not only the canonical chain but also any dangling side chains.

**Comment by meowsbits** [comment]
Just bumping. This has happened to me twice now.

**Comment by karalabe** [comment]
Fixed by https://github.com/ethereum/go-ethereum/pull/21409.
